statshouse.hpp

```diff
@@ -937,25 +964,88 @@ class Registry {
 			, timestamp{timestamp}
 			, last_flush_timestamp{0}
 			, regular{0}
+			, last{0}
 			, rand{}
 			, queue_ptr{nullptr} {
 		}
 		TransportUDP::MetricBuilder key;
 		uint32_t timestamp;
 		uint32_t last_flush_timestamp;
 		int regular;
+		int last;
 		PRNG rand;
 		multivalue value;
 		std::mutex mu;
 		std::shared_ptr<bucket> *queue_ptr;
 	};
-public:
-	class MetricRef {
-	public:
-		MetricRef(Registry *registry, const TransportUDP::MetricBuilder &key)
+	struct bucket_ref {
+		bucket_ref()
+			: registry{nullptr} {
+		}
+		bucket_ref(Registry *registry, const TransportUDP::MetricBuilder &key)
 			: registry{registry}
 			, ptr{registry->get_or_create_bucket(key)} {
 		}
+		Registry *registry;
+		std::shared_ptr<bucket> ptr;
+	};
+	struct bucket_regular_ref : bucket_ref {
+		bucket_regular_ref()
+			: last{false} {
+		}
+		bucket_regular_ref(Registry *registry, const TransportUDP::MetricBuilder &key, bool last)
+			: bucket_ref{registry, key}
+			, last{last} {
+			init();
+		}
+		~bucket_regular_ref() {
+			if (!ptr) {
+				// might be null after move constructor/assignment
+				return;
+			}
+			auto n = 0, m = 0;
+			{
+				std::lock_guard<std::mutex> bucket_lock{ptr->mu};
+				n = --ptr->regular;
+				if (last) {
+					m = --ptr->last;
+				}
+			}
+			if (n < 0 || m < 0) {
+				registry->log_error("unexpected bucket regular reference count value %d, %d", n, m);
+			}
+		}
+		bucket_regular_ref(const bucket_regular_ref &other)
+			: bucket_ref{other}
+			, last{other.last} {
+			if (ptr) {
+				init();
+			}
+		}
+		bucket_regular_ref &operator=(bucket_regular_ref other) {
```

Comment:
It's some weird flex that I cannot comprehend. I might be wrong but looks like copy assignment should be here.